safe/impact_functions/inundation/flood_raster_road/impact_function.py

```diff
@@ -0,0 +1,394 @@
+from safe.common.tables import Table, TableRow
+from safe.impact_functions.inundation.flood_raster_road\
+    .metadata_definitions import FloodRasterRoadsMetadata
+from safe.utilities.i18n import tr
+from safe.storage.vector import Vector
+from safe.common.utilities import get_utm_epsg, unique_filename
+from safe.common.exceptions import GetDataError
+from safe.gis.qgis_raster_tools import clip_raster
+from safe.gis.qgis_vector_tools import (
+    extent_to_geo_array,
+    create_layer)
+
+
+def _raster_to_vector_cells(raster, threshold_min, threshold_max, roads_crs):
+    """ take raster and generate vectors features (rectangles) for cells
+        which are within threshold (thr_min < V < thr_max) """
+
+    # get raster data
+    provider = raster.dataProvider()
+    extent = provider.extent()
+    raster_cols = provider.xSize()
+    raster_rows = provider.ySize()
+    block = provider.block(1, extent, raster_cols, raster_rows)
+    raster_xmin = extent.xMinimum()
+    raster_ymax = extent.yMaximum()
+    cell_width = extent.width() / raster_cols
+    cell_height = extent.height() / raster_rows
+
+    uri = "Polygon?crs=" + roads_crs.authid()
+    vl = QgsVectorLayer(uri, "flood polygons", "memory")
+    features = []
+
+    # prepare coordinate transform to reprojection
+    ct = QgsCoordinateTransform(raster.crs(), roads_crs)
+
+    for y in xrange(raster_rows):
+        for x in xrange(raster_cols):
+            # only use cells that are within the specified threshold
+            value = block.value(y, x)
+            if value < threshold_min or value > threshold_max:
+                continue
+
+            # construct rectangular polygon feature for the cell
+            x0 = raster_xmin+(x*cell_width)
+            x1 = raster_xmin+((x+1)*cell_width)
+            y0 = raster_ymax-(y*cell_height)
+            y1 = raster_ymax-((y+1)*cell_height)
+            outer_ring = [QgsPoint(x0, y0), QgsPoint(x1, y0),
+                          QgsPoint(x1, y1), QgsPoint(x0, y1),
+                          QgsPoint(x0, y0)]
+            geometry = QgsGeometry.fromPolygon([outer_ring])
+            geometry.transform(ct)
+            f = QgsFeature()
+            f.setGeometry(geometry)
+            features.append(f)
+
+    _, features = vl.dataProvider().addFeatures(features)
+
+    # construct a temporary map for fast access to features by their IDs
+    # (we will be getting feature IDs from spatial index)
+    flood_cells_map = {}
+    for f in features:
+        flood_cells_map[f.id()] = f
+
+    # build a spatial index so we can quickly identify
+    # flood cells overlapping roads
+    if QGis.QGIS_VERSION_INT >= 20800:
+        # woohoo we can use bulk insert (much faster)
+        index = QgsSpatialIndex(vl.getFeatures())
+    else:
+        index = QgsSpatialIndex()
+        for f in vl.getFeatures():
+            index.insertFeature(f)
+
+    return index, flood_cells_map
+
+
+def _add_output_feature(features, geom, is_flooded,
```

Comment:
Can you make a unit test for this please? I also think we should move this to one of our utility classes e.g. safe/gis/qgis_vector_tools.py